### PR TITLE
Smart Query - Adds the possibility to use index hints

### PR DESCRIPTION
**Pull request by oolscireum**
Adds the possibility to use index hints like "USE INDEX", "FORCE INDEX" or "IGNORE INDEX" in the Smart Query.
This enables the option to add hints for a query, when the db optimizer is using sub optimal indexes or ignores specific indexes.

Fixes: SIRI-947

### CODE REVIEW

**Review by fhaScireum — APPROVED** [review]

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 1]
```diff
@@ -73,6 +73,14 @@ public class SmartQuery<E extends SQLEntity> extends Query<SmartQuery<E>, E, SQL
     protected List<String> groupBys = Collections.emptyList();
     protected List<SQLConstraint> constraints = new ArrayList<>();
     protected Database db;
+    protected Map<String, String> indexHints = new HashMap<>();
+
+    /**
+     * Indicates the "FOR" hint for the optimizer, when used with "USE INDEX" and "IGNORE INDEX".
+     */
+    public enum IndexForHint {
+        JOIN, ORDER_BY, GROUP_BY
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:82]: **Optional:** Short JavaDoc explaining what each hint does?

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 2]
```diff
@@ -105,6 +113,91 @@ public SmartQuery<E> where(SQLConstraint constraint) {
         return this;
     }
 
+    /**
+     * Adds a "USE INDEX" hint for a table that recommends an index to te optimizer.
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:117]:
```suggestion
     * Adds a "USE INDEX" hint for a table that recommends an index to the optimizer.
```

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 3]
```diff
@@ -105,6 +113,91 @@ public SmartQuery<E> where(SQLConstraint constraint) {
...
+     * @param indexName the name of the index to use.
+     * @param <T>       the type of the entities being queried.
+     * @return the query itself for fluent method calls.
+     */
+    public <T extends SQLEntity> SmartQuery<E> useIndex(Class<T> table, String indexName) {
+        return useIndex(table, null, indexName);
+    }
+
+    /**
+     * Adds a "USE INDEX" hint for a table that recommends an index to te optimizer.
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:129]:
```suggestion
     * Adds a "USE INDEX" hint for a table that recommends an index to the optimizer.
```

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 4]
```diff
@@ -105,6 +113,91 @@ public SmartQuery<E> where(SQLConstraint constraint) {
...
+     *
+     * @param table     the table for which the index should be used.
+     * @param forHint   the action for which the index should be used. Can be null.
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:132]: **Convention:** Nullability shouldn't be declared in JavaDoc but via the correct annotations.

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 5]
```diff
@@ -105,6 +113,91 @@ public SmartQuery<E> where(SQLConstraint constraint) {
...
+     * @param <T>       the type of the entities being queried.
+     * @return the query itself for fluent method calls.
+     */
+    public <T extends SQLEntity> SmartQuery<E> useIndex(Class<T> table, IndexForHint forHint, String indexName) {
+        indexHint(table, "USE", forHint, indexName);
+        return this;
+    }
+
+    /**
+     * Adds a "FORCE INDEX" hint for a table that forces an index to te optimizer.
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:143]:
```suggestion
     * Adds a "FORCE INDEX" hint for a table that forces an index to the optimizer.
```
Is your IntellIJ spell checker disabled? 😇

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 6]
```diff
@@ -105,6 +113,91 @@ public SmartQuery<E> where(SQLConstraint constraint) {
...
+
+    /**
+     * Adds a "IGNORE INDEX" hint for a table that makes the optimizer to not consider the given index.
+     *
+     * @param table     the table for which the index should be used.
+     * @param indexName the name of the index to use.
+     * @param <T>       the type of the entities being queried.
+     * @return the query itself for fluent method calls
+     */
+    public <T extends SQLEntity> SmartQuery<E> ignoreIndex(Class<T> table, String indexName) {
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:163]: **Question:** Do we really need/want all these overloads just to hide a single nullable parameter?

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 7]
```diff
@@ -105,6 +113,91 @@ public SmartQuery<E> where(SQLConstraint constraint) {
...
+    public <T extends SQLEntity> SmartQuery<E> ignoreIndex(Class<T> table, IndexForHint forHint, String indexName) {
+        indexHint(table, "IGNORE", forHint, indexName);
+        return this;
+    }
+
+    /**
+     * Adds a hint for a table that is used by the optimizer.
+     *
+     * @param table     the table for which the index should be used.
+     * @param type      the type of the hint. Can be "USE", "FORCE" or "IGNORE".
```
Comment (sabieber) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:185]: **Optional:** Better to link to the enum here or via `@see`

File: src/main/java/sirius/db/jdbc/SmartQuery.java (hunk 5 above)
Comment (oolscireum) [reply src/main/java/sirius/db/jdbc/SmartQuery.java:143]: Looks like it 😲

File: src/main/java/sirius/db/jdbc/SmartQuery.java (hunk 6 above)
Comment (oolscireum) [reply src/main/java/sirius/db/jdbc/SmartQuery.java:163]: I removed the overloads for the USE and IGNORE methods. The other ones aren't just hiding a nullable.

**Review by sabieber — APPROVED** [review]

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 8]
```diff
@@ -105,6 +124,67 @@ public SmartQuery<E> where(SQLConstraint constraint) {
         return this;
     }
 
+    /**
+     * Adds a "USE INDEX" hint for a table that recommends an index to te optimizer.
```
Comment (mkeckmkeck) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:128]:
```suggestion
     * Adds a "USE INDEX" hint for a table that recommends an index to the optimizer.
```

**Review by mkeckmkeck — APPROVED** [review]
Please make the PR mergable.

**Review by mko-sci — APPROVED** [review]

**Pull request merged**